src/fu-engine.c

```diff
@@ -852,29 +852,102 @@ fu_engine_validate_bios_attr_input(FwupdBiosAttr *attr, const gchar **value, GEr
 	return TRUE;
 }
 
+static gboolean
+fu_engine_modify_single_bios_attr(FuEngine *self,
+				  const gchar *key,
+				  const gchar *value,
+				  GError **error)
+{
+	FwupdBiosAttr *attr = fu_context_get_bios_attr(self->ctx, key);
+	const gchar *tmp = value;
+
+	if (!fu_engine_validate_bios_attr_input(attr, &tmp, error))
+		return FALSE;
+
+	if (!fu_engine_update_bios_attr(attr, tmp, error))
+		return FALSE;
+
+	return TRUE;
+}
+
+static gboolean
+fu_engine_modify_json_bios_attr(FuEngine *self, const gchar *blob, GError **error)
+{
+	JsonObject *obj;
+	JsonNode *node;
+	g_autoptr(JsonParser) parser = json_parser_new();
+	g_autoptr(FuBiosAttrs) new_bios_attrs = fu_bios_attrs_new();
+	g_autoptr(GPtrArray) new_items = NULL;
+
+	if (!json_parser_load_from_data(parser, blob, -1, error))
+		return FALSE;
+	node = json_parser_get_root(parser);
+	obj = json_node_get_object(node);
+	if (!json_object_has_member(obj, "BiosAttributes")) {
+		g_set_error_literal(error,
+				    G_IO_ERROR,
+				    G_IO_ERROR_INVALID_DATA,
+				    "missing BiosAttributes");
+		return FALSE;
+	}
+	if (!fu_bios_attrs_from_json(new_bios_attrs, node, error))
+		return FALSE;
+
+	new_items = fu_bios_attrs_get_all(new_bios_attrs);
+	for (guint i = 0; i < new_items->len; i++) {
+		FwupdBiosAttr *item_attr = g_ptr_array_index(new_items, i);
+		FwupdBiosAttr *ctx_attr =
+		    fu_context_get_bios_attr(self->ctx, fwupd_bios_attr_get_id(item_attr));
+		if (!ctx_attr) {
+			g_debug("%s doesn't exist on this system, skipping",
+				fwupd_bios_attr_get_id(item_attr));
+			continue;
+		}
+		if (g_strcmp0(fwupd_bios_attr_get_current_value(ctx_attr),
+			      fwupd_bios_attr_get_current_value(item_attr)) == 0) {
+			g_debug("%s is already set to %s",
+				fwupd_bios_attr_get_id(item_attr),
+				fwupd_bios_attr_get_current_value(item_attr));
+			continue;
+		}
+		if (!fu_engine_modify_single_bios_attr(self,
+						       fwupd_bios_attr_get_id(item_attr),
+						       fwupd_bios_attr_get_current_value(item_attr),
+						       error))
+			return FALSE;
+	}
+
+	return TRUE;
+}
+
 /**
  * fu_engine_modify_bios_attr:
  * @self: a #FuEngine
- * @key: the name of the BIOS attribute, e.g. `SleepMode`
- * @value: the value to set, e.g. 'S3'
+ * @key: "json" or the name of the BIOS attribute, e.g. `SleepMode`
+ * @value: contents of JSON payload OR the value to set, e.g. 'S3'
  * @error: (nullable): optional return location for an error
  *
  * Use the kernel API to set a BIOS attribute.
+ * - This can be done with a key/value pair
+ *   or
+ * - This can be done with key set to "json" and value being a JSON blob
  *
  * Returns: %TRUE for success
  **/
 gboolean
 fu_engine_modify_bios_attr(FuEngine *self, const gchar *key, const gchar *value, GError **error)
```

Comment:
not convinced we want to handle this inside this function -- maybe create a fu_engine_modify_bios_attrs and handle the "json" key in the engine/tool?

Reply:
How about we just leave all the JSON parsing to the tool and send a set of pairs every time?  Then we don't have a special "json" key.

Reply:
I think that's even better, although isn't "atomic" -- which is probably fine. If we need atomic we can have a different method call.